First and second chapter of the training

artn-odoo opened the pull request.

artn-odoo committed:
First changes

artn-odoo committed:
[IMP] change the commits name to fit company regulations

artn-odoo committed:
[REF] refactors the commit messages

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/230.png)](https://mergebot.odoo.com/odoo/tutorials/pull/230)

ethanrobv commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'demo': [
+        'demo/demo_data.xml',
+    ],
+}
```
You generally (actually, pretty much always) want to leave an empty newline at the end of a file

artn-odoo committed:
[IMP] Adds the model class. The class only has a name for now, it is in the models folder, where all models will be.

artn-odoo committed:
[IMP] Adds the model class. The class only has a name for now, it is in the models folder, where all models will be.

artn-odoo committed:
[IMP] adds the base for the estate module such as the name and the description. Can't see the database.

artn-odoo committed:
[IMP] Introducing security and bug fixing. The init file was missing an underscore, which made the import useless. The security rule is now enabled and can be viewed in ir.model.access.csv in the security folder.

nesma-neha commented on estate/data/estate.property.csv:
```diff
@@ -0,0 +1 @@
+"id","name"
```
You don't need this file

nesma-neha commented on estate/models/estate_property.py:
```diff
@@ -1,6 +1,6 @@
 from odoo import models, fields
 
-class EstateProperty(models.Model):
+class estate_property(models.Model):
```
`class EstateProperty(models.Model):`

Python files should be in Pascal Case

nesma-neha commented on estate/__init__.py:
```diff
@@ -1 +1,2 @@
-from . import models
\ No newline at end of file
+from . import models
+from . import security
```
Data files shouldn't be imported in the __init__.py
You only import folders that contain python files

nesma-neha reviewed (commented):
The commit title is the first line of the commit message. 
then you should add a new line to write the commit description.

artn-odoo committed:
[IMP] Adding some views.
> Add a form view and a list view for the estate_property model. Also adds the action and some menu items. Refactors the model to add the needed attributes.

nesma-neha commented on estate/models/estate_property.py:
```diff
@@ -1,8 +1,31 @@
+
+
+
+
```
remove extra lines

nesma-neha commented on estate/views/estate_menu.xml:
```diff
@@ -0,0 +1,10 @@
+            </menuitem>
+        </menuitem>
+    </data>
+</odoo>
```
add line at EOF

artn-odoo committed:
[IMP] Adding views to the module.
A custom form and list view has been added. A custom search for almost all fields is available. A custom filter to filter out certain states has been implemented as well as a group by option on the postcode.Various view ameliorations.  Various upgrades aswell in concerning the previous PR returns.

artn-odoo committed:
[IMP] Adds a lot of menu items.
These menu items can be used to give more informations about the estate properties that has to be sold.
Doesn't work as there is a bug on the Offer part of the project

artn-odoo committed:
[FIX] Fix the order of the file and end of chapter 7.
The structure of the XML files was not how it should have been. Some views were called before being declared thus resulting in a bug. The chapter 7 has been completed.

artn-odoo committed:
[IMP] add computed and onchange fields.
The computed fields added are used to know the total area of a property and the onchange field is used to change the garden area and orientation once the checkbox is checked.
The best offer for a property is also added and it changes automatically when a new best offer is created

artn-odoo committed:
[IMP] Add actions to the views
The actions added are buttons that can change the state of models when clicked. It can change a house to the cancelled or sold status. They can accept or reject an offer and change the corresponding values in the models, like the selling price or the buyer

artn-odoo committed:
[ADD] add gitignore for launch.json

artn-odoo committed:
[IMP] Add constraints to various models.
There are SQL constraints to make sure that tags are unique, aswell as making sure that every prices for a property is strictly positive. There is also one Python constraint to make sure that the selling price for a property is never below 90% of the expected price of that same property.

ethanrobv commented on estate/models/estate_property.py:
```diff
@@ -62,3 +62,24 @@ def _onchange_garden(self):
             self.garden_area = 0
             self.garden_orientation = False
+
+    def cancel_property(self):
```
Leave a newline betwixt method definitions

ethanrobv committed:
[ADD] created probably conflict in ResUsers extension

resolve it

artn-odoo committed:
[IMP] Add all kind of QoL.
Add offers related to a type on the type form view. Add a property state as a widget on top of the form. Add an order to all models in the list view. Add a handle to change this order. Add colors for tags. Add invisible for some functionnalities if conditions are met. Add an offer as readonly on certain condition. The list view is now editable. Make date availability an optionnal in the list view. Add some decoration so it is clearer without needing to display the status. The available filter is enabled by default. Add a search domain on name and description. Add a stat button to get the number of offers related to a type. When clicked, the list view of only those offers is displayed.

ethanrobv commented on estate/models/estate_property_offer.py:
```diff
@@ -16,6 +17,13 @@ class EstatePropertyOffer(models.Model):
     property_id = fields.Many2one("estate.property", required=True)
     validity = fields.Integer(name="Validity", default=7)
     date_deadline = fields.Date(name="Deadline", compute="_compute_deadline", inverse="_inverse_validity")
+    #property_type_id = fields.Many2one("estate.property.types", name="Property type")
```
remove

ethanrobv commented on estate/models/estate_property_offer.py:
```diff
@@ -16,6 +17,13 @@ class EstatePropertyOffer(models.Model):
     date_deadline = fields.Date(name="Deadline", compute="_compute_deadline", inverse="_inverse_validity")
+    #property_type_id = fields.Many2one("estate.property.types", name="Property type")
+    property_type_id = fields.Many2one(
+        "estate.property.types", 
```
You don't have to change it all at this point, but we prefer single quotes in python files unless using double is required for some semantic reasons

ethanrobv commented on estate/models/estate_property_offer.py:
```diff
@@ -32,13 +40,16 @@ def _compute_deadline(self):
-            if isinstance(record.create_date, fields.Date):
-                record.validity = (record.date_deadline - record.create_date).days
+            if record.date_deadline:
+                # Convert create_date to a date object if it exists
```
imo it doesn't need an explanation

ethanrobv commented on estate/views/estate_property_offer_view.xml:
```diff
@@ -22,6 +23,13 @@
             <field name="view_id" ref="view_estate_property_offer_list"/>
         </record>
+        <record id="action_estate_property_offer_from_types" model="ir.actions.act_window">
+            <field name="name">OffersFromTypes</field>
```
You can either make it regularly spaced, i.e., `Offers from types` (or whatever the name)
or you can make it snake case `offers_from_types`

ethanrobv commented on estate/views/estate_property_view.xml:
```diff
@@ -94,6 +96,7 @@
                     <field name="living_area" string="Living area (m²)"/>
                     <field name="type_id" string="Property type"/>
                     <filter name="available" domain="['|', ('state', '=', 'new'), ('state', '=', 'offer received')]"/>
+                    <field name="description" string="Name and description" filter_domain="['|', ('name', 'ilike', self), ('description', 'ilike', self)]"/>
```
You can check equality with strings, `ilike` is really more for when you need wildcards/only know part of the string you are trying to match

artn-odoo committed:
[IMP] Add a res users field and constraints on create and delete.
Add a constraint on creating an offer so that the price can't be lower than the best price for that property. A warning remains in this.  Add one on delete where the state can only be new or cancelled, otherwise the record won't be deleted. Extend the res.users model to add a field displaying the properties.

artn-odoo committed:
[ADD] Add the new estate_account module.
This is made to automatically create invoices once the house is sold. Doesn't work for now as the override of the sold_property function doesn't work.

artn-odoo committed:
[IMP] Bug solving and add kanban view.
Adds a kanban view for the estate properties, grouped by type and not draggable. The informations are not yet displayed on it.

artn-odoo committed:
[FIX] Fixed bug in kanban view.
The fields were not correctly formated and some didn't appear. They did not appear because the field state was not declared before the template so we couldn't access it's raw value.

ethanrobv commented:
You can interactive rebase (`-i`) with `--root` to change your first commit header (which is currently causing the [style check](https://runbot.odoo.com/runbot/build/74202403) to fail)

artn-odoo committed:
[FIX] Fix warnings and bugs on the new module.
The lingering warning has been fixed by replace the decorator on the create override method. The bug preventing the sold_property method on the new module to override the main one was caused by the model from the new module having a name. It has been fixed.
